### PR TITLE
feat : Viewmodel for DayScheduleFragment

### LINKED ISSUE #2031: Create ViewModel for the DayScheduleFragment
**Actual Behaviour**

There is no ViewModel for the DayScheduleFragment

**Expected Behaviour**

There should be one.

**Steps to reproduce it**

NA

**LogCat for the issue**

NA

**Screenshots of the issue**

NA

**Would you like to work on the issue?**

Yes!

**Target Milestone to solve the issue**

NA

**Other information**

Parent issue #2015

### PR DESCRIPTION
**Pull request by samagra14**
Fixes #2031 

Changes: Adds viewModel for DayScheduleFragments

Screenshots for the change: NA

@iamareebjamal @srv-twry Please review the changes. I have added some getters and setters in other classes as they were giving problems with the build. I feel that the code in filter method is not optimal. I will improve this method in a new issue or it can be improved once we shift to Room. I have tested the code on MiA1 Android 8.0 Oreo. It is working fine.

### CODE REVIEW

File: android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java [hunk 1]
```diff
@@ -82,37 +89,41 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
 
         Utils.registerIfUrlValid(swipeRefreshLayout, this, this::refresh);
         setUpRecyclerView();
+        Log.v(LOG_TAG,"Fragment Opened");
```
Comment (dr0pdb) [inline android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java]: Please remove log statements

File: android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java [hunk 2]
```diff
@@ -49,26 +52,30 @@
  */
 public class DayScheduleFragment extends BaseFragment implements SearchView.OnQueryTextListener {
 
-    final private String SEARCH = "searchText";
 
-    private String searchText = "";
+
+    private String searchText ;
```
Comment (dr0pdb) [inline android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java]: Remove space before semi-colon

File: android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java [hunk 3]
```diff
@@ -49,26 +52,30 @@
...
     @BindView(R.id.schedule_swipe_refresh) SwipeRefreshLayout swipeRefreshLayout;
     @BindView(R.id.list_schedule) RecyclerView dayRecyclerView;
     @BindView(R.id.txt_no_schedule) TextView noSchedule;
     @BindView(R.id.txt_no_result_schedule) protected TextView noResultsSchedule;
 
-    private List<Session> sessions = new ArrayList<>();
-    private List<Session> filteredSessions = new ArrayList<>();
+    private List<Session> sessionsList = new ArrayList<>();
+    private List<Session> filteredSessionsList = new ArrayList<>();
+
```
Comment (dr0pdb) [inline android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java]: Please remove the extra line

File: android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        this.date = date;
+    }
+
+    public LiveData<RealmResults<Session>> getRealmResults() {
+        if (realmResults == null)
+            realmResults = new MutableLiveData<>();
+        realmResults.setValue(realmRepo.getSessionsByDate(date, SortOrder.sortTypeSchedule()));
+        return realmResults;
+    }
+
```
Comment (dr0pdb) [inline android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java:44]: Why so many extra lines ?

**Review by dr0pdb — CHANGES_REQUESTED** [review]
There are way too many formatting errors. I have pointed some of them out. Please find all of them and fix them, then only the reviewers can review the actual refactor.

File: android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java (hunk 4 above)
Comment (samagra14) [reply android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java:44]: Checked.

File: android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java (hunk 2 above)
Comment (samagra14) [reply android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java]: Checked.

File: android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java (hunk 3 above)
Comment (samagra14) [reply android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java]: Checked.

File: android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java [hunk 5]
```diff
@@ -53,16 +52,17 @@
 public class DayScheduleFragment extends BaseFragment implements SearchView.OnQueryTextListener {
 
 
-
-    private String searchText ;
-    private SearchView searchView;
     private static final String LOG_TAG = "DayScheduleFragment";
-
-    @BindView(R.id.schedule_swipe_refresh) SwipeRefreshLayout swipeRefreshLayout;
```
Comment (dr0pdb) [inline android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java]: Why have you changed this ? It was good before.

File: android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java [hunk 6]
```diff
@@ -83,36 +88,37 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle sa
...
-
-        realmResults = realmRepo.getSessionsByDate(date, SortOrder.sortTypeSchedule());
-        realmResults.addChangeListener((sortedSessions, orderedCollectionChangeSet) -> {
-            sessions.clear();
-            sessions.addAll(sortedSessions);
-
-            filteredSessions.clear();
-            filteredSessions.addAll(sortedSessions);
+        dayScheduleFragmentViewModel = ViewModelProviders.of(this)
+                .get(DayScheduleFragmentViewModel.class);
```
Comment (dr0pdb) [inline android/app/src/main/java/org/fossasia/openevent/fragments/DayScheduleFragment.java:92]: No need to add it to a new line. We don't need to strictly follow the 100 character limit, if it is not affecting the readability of the code i.e. user doesn't need to scroll left/right.

File: android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+import android.arch.lifecycle.ViewModel;
+
+import org.fossasia.openevent.data.Session;
+import org.fossasia.openevent.dbutils.RealmDataRepository;
+import org.fossasia.openevent.utils.SortOrder;
+
+import io.realm.RealmResults;
+
+/**
+ * Created by samagra on 23/1/18.
```
Comment (dr0pdb) [inline android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java:14]: Remove this template.

File: android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ */
+
+public class DayScheduleFragmentViewModel extends ViewModel {
+    private MutableLiveData<RealmResults<Session>> realmResults;
+    private String searchText = "";
+
+
```
Comment (dr0pdb) [inline android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java:21]: Remove this extra line.

File: android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    private RealmDataRepository realmRepo;
+    private String date;
+
+    public DayScheduleFragmentViewModel() {
+        realmRepo = RealmDataRepository.getDefaultInstance();
+
```
Comment (dr0pdb) [inline android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java:27]: Remove this extra line.

File: android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    public String getSearchText() {
+        return searchText;
+    }
+
+    public void setSearchText(String searchText) {
+        this.searchText = searchText;
+    }
+
+    @Override
+    protected void onCleared() {
```
Comment (dr0pdb) [inline android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java:54]: If you're just calling the super class then this method is redundant and can be removed.

**Review by dr0pdb — CHANGES_REQUESTED** [review]
In my last review, i pointed out a few formatting errors. Those aren't the only formatting errors in the PR. Please find the others and fix them. 

Moreover, I would advise you to look at the implementation of other ViewModels to see how we are implementing them.

You have made a huge amount of redundant changes in  the data module, please explain their usage. I have myself added a large number of ViewModels and never had to make any changes to the repository layer.

Please check other ViewModels to see how we're implementing the filtering in the ViewModel only.

Most of the bussiness logic in your PR is still inside the Fragment, so kindly move them to the ViewModel. The key is to make the Activity as dumb as possible.

File: android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java (hunk 8 above)
Comment (samagra14) [reply android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java:21]: What is wrong with blank lines. Already, the documentation is poor. There are no Javadoc comments. In such cases, blank lines can actually point out changes in context.

**Comment by samagra14** [comment]
@srv-twry I have not moved the filtering code to the ViewModel because as I said earlier, it can be a lot more optimized. It just repeatedly recreates filtered lists. Anyways, I will move it to the ViewModel.

File: android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java (hunk 8 above)
Comment (dr0pdb) [reply android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java:21]: I never said adding blank lines is wrong. Adding two/three of them where you should at max. add one line is wrong in my opinion. A single blank line can point out changes in context.

>  Already, the documentation is poor. There are no Javadoc comments.

**Then please help us make it better.** 
Please create a new issue if you think it should be present and maintainers will surely respond.

**Comment by dr0pdb** [comment]
@samagra14 
> it can be a lot more optimized.

Sure, looking forwards to your point of view. 

> Anyways, I will move it to the ViewModel.

If you think you've a better idea, then please implement that in this PR only, since the addition of ViewModel also includes moving the filtering logic from Adapter to ViewModel.

File: android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java (hunk 8 above)
Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/viewmodels/DayScheduleFragmentViewModel.java:21]: Only one blank line is standard according to any Java-style guideline @samagra14

**Comment by iamareebjamal** [comment]
@samagra14 If it can be optimized, we'll surely like to see the proposal but it would mean that the same can be done with other ViewModels as well. So it needs its own issue. We can follow the same pattern here as with other ViewModels and then handle the optimization as a new issue in separate PR

Leaving the state hanging in between not implemented or different from other components poses this problem -> Maybe no one takes a new issue up to fix this incomplete ViewModel, or no one takes up the issue of optimizing all ViewModels. This resolves to incomplete refactoring of project leading to separate architecture in components making it hard for beginners to understand (as you can already see)

**Comment by samagra14** [comment]
I am sorry for replying late. I have been busy with the republic day prep. I will make the changes by Saturday. Do I need to squash my commits?

**Comment by iamareebjamal** [comment]
Yes

**Comment by iamareebjamal** [comment]
@srv-twry @mayank8318 @nikit19 Please review and ensure that opened PRs are final and ready. Thanks :)

**Comment by dr0pdb** [comment]
@samagra14 You haven't made any changes. Please update the PR with the requested changes :)

**Comment by samagra14** [comment]
@srv-twry I am sorry. I have been busy lately. I will update the PR with requested changes positively within a couple of days.

**Comment by dr0pdb** [comment]
@samagra14 Status ? 
If you don't update this PR within the next couple of days, i will myself take this issue and create one since it is stalling the progress in other issues. So please update the PR. Thanks.

**Comment by AkshayCHD** [comment]
@iamareebjamal @srv-twry can i work on this issue

**Comment by dr0pdb** [comment]
@AkshayCHD There is another active PR on this.

**Comment by dr0pdb** [comment]
@iamareebjamal Please close this since another PR has been merged solving the issue.

**Pull request closed without merging**